Convert HPA controller to support HPA v2 mechanics

DirectXMan12 opened the pull request:
This PR converts the HPA controller to support the mechanics from HPA v2.
The HPA controller continues to make use of the HPA v1 client, but utilizes
the conversion logic to work with autoscaling/v2alpha1 objects internally.

It is the follow-up PR to #36033 and part of kubernetes/features#117.

**Release note**:
```release-note
NONE
```

DirectXMan12 committed:
Convert HPA controller to use autoscaling/v2alpha1

This commit converts the HPA controller over to using the new version of
the HorizontalPodAutoscaler object found in autoscaling/v2alpha1.  Note
that while the autoscaler will accept requests for object metrics, the
scale client will return an error on attempts to get object metrics
(since that requires the new custom metrics API, which is not yet
implemented).

This also enables the HPA object in v2alpha1 as a retrievable API
version by default.

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/41272)
<!-- Reviewable:end -->

DirectXMan12 commented:
cc @jszczepkowski

DirectXMan12 commented:
cc @caesarxuchao @deads2k @liggitt @lavalamp re the discussion the other day about different object versions in the controller on Slack, hopefully this solution doesn't make you recoil in horror :-P

jszczepkowski commented on pkg/controller/podautoscaler/metrics/utilization.go:
```diff
@@ -23,9 +23,10 @@ import (
 // and a target utilization percentage, and calcuates the the ratio of
 // desired to actual utilization (returning that and the actual utilization)
-func GetResourceUtilizationRatio(metrics PodResourceInfo, requests map[string]int64, targetUtilization int32) (float64, int32, error) {
+func GetResourceUtilizationRatio(metrics PodMetricsInfo, requests map[string]int64, targetUtilization int32) (float64, int32, int64, error) {
```
Can you please name the returned values? Now it is unclear which one is which.

jszczepkowski replied:
Also the comment should be updated.

jszczepkowski commented on pkg/controller/podautoscaler/metrics/utilization.go:
```diff
@@ -36,29 +37,43 @@ func GetResourceUtilizationRatio(metrics PodResourceInfo, requests map[string]in
 // (returning that and the actual utilization)
-func GetMetricUtilizationRatio(metrics PodMetricsInfo, targetUtilization float64) (float64, float64) {
-	metricsTotal := float64(0)
+func GetMetricUtilizationRatio(metrics PodMetricsInfo, targetUtilization int64) (float64, int64) {
```
Can you name the returned values?

jszczepkowski commented on pkg/controller/podautoscaler/replica_calculator.go:
```diff
@@ -149,12 +243,13 @@ func (c *ReplicaCalculator) GetResourceReplicas(currentReplicas int32, targetUti
+// GetMetricReplicas calculates the desired replica count based on a target metric utilization
+// (as a milli-value) for pods matching the given selector in the given namespace, and the
+// current replica count
+func (c *ReplicaCalculator) GetMetricReplicas(currentReplicas int32, targetUtilization int64, metricName string, namespace string, selector labels.Selector) (replicaCount int32, utilization int64, timestamp time.Time, err error) {
```
It seems that methods GetMetricReplicas and GetRawResourceReplicas are similar. Did you consider making their implementation to use a common method?

jszczepkowski commented on pkg/controller/podautoscaler/replica_calculator.go:
```diff
@@ -245,3 +337,20 @@ func (c *ReplicaCalculator) GetMetricReplicas(currentReplicas int32, targetUtili
+
+// GetObjectMetricReplicas calculates the desired replica count based on a target metric utilization (as a milli-value)
+// for the given object in the given namespace, and the current replica count.
+func (c *ReplicaCalculator) GetObjectMetricReplicas(currentReplicas int32, targetUtilization int64, metricName string, namespace string, objectRef *autoscaling.CrossVersionObjectReference) (replicaCount int32, utilization int64, timestamp time.Time, err error) {
```
Do you have a unittest for this method?

DirectXMan12 replied:
there isn't one because the existing replica calc test is structured around the actual metrics client, so it needs a working implementation of `GetObjectMetrics`, which we don't have (it just returns a "not-implemented" error).  I can either a) restructure the entire test, or b) add in the test in part 3 of this PR, which adds support for custom metrics using the custom metrics API types.

jszczepkowski commented on pkg/controller/podautoscaler/horizontal_test.go:
```diff
@@ -704,20 +683,6 @@ func TestScaleUpCMUnreadyNoScaleWouldScaleDown(t *testing.T) {
 	tc.runTest(t)
 }
 
-func TestDefaultScaleDown(t *testing.T) {
```
Why was this test removed?

DirectXMan12 replied:
We don't have implicit defaulting any more -- it's done a the API level (defaulting for v2alpha1, and conversion for v1 for reasons of method signatures and JSON unmarshalling).

jszczepkowski commented on pkg/controller/podautoscaler/horizontal_test.go:
```diff
@@ -504,58 +526,6 @@ func (tc *testCase) runTest(t *testing.T) {
 	tc.verifyResults(t)
 }
 
-func TestDefaultScaleUpRC(t *testing.T) {
```
Why were these tests removed?

jszczepkowski commented on pkg/controller/podautoscaler/horizontal.go:
```diff
@@ -139,141 +150,140 @@ func (a *HorizontalController) Run(stopCh <-chan struct{}) {
 }
 
-// getLastScaleTime returns the hpa's last scale time or the hpa's creation time if the last scale time is nil.
-func getLastScaleTime(hpa *autoscaling.HorizontalPodAutoscaler) time.Time {
```
Don't we need it anymore? Where is lastScaleTime set to CreationTimestamp?

DirectXMan12 replied:
don't think this was intentional -- probably got lost amongst the rebases

DirectXMan12 replied:
Ah, I know why this got removed -- it's largely irrelevent with the improved handling for missing pods -- under the new scheme, we'll only error out if no metrics *at all* are present, which will only occur a) right after the deployment is initially created, or b) if we have an actual error.

jszczepkowski commented on pkg/controller/podautoscaler/horizontal.go:
```diff
@@ -139,141 +150,140 @@ func (a *HorizontalController) Run(stopCh <-chan struct{}) {
 
-	for _, customMetricTarget := range targetList.Items {
+	for i, metricSpec := range metricSpecs {
 		if scale.Status.Selector == nil {
```
I don't understand this check: scale.Status.Selector is array, we should use len to get its length. I guess this check will be always false.

DirectXMan12 replied:
selector is a map, and technically maps can be nil.  But I think this was from when we were using the internal version, and never got properly updated.  I'll fix it.

DirectXMan12 commented:
@jszczepkowski I believe I've address all of your comments (either as code changes or as replies).  PTAL.

jszczepkowski commented:
/lgtm /approve

Please, fix e2e tests.

DirectXMan12 commented:
@k8s-bot gce etcd3 e2e test this

Unsure if this was a flake, but looks like it might be.  Let's try again...

DirectXMan12 commented:
@k8s-bot cvm gke e2e test this

@k8s-bot kubemark e2e test this

looks like the first one was a flake, let's try the rest again

DirectXMan12 committed:
Fix HPA v1 Conversion Bug

There was a bug in the HPA v1 conversion logic that would occur when
a custom metric and a metric that was encoded in v1 as
targetCPUUtilizationPercentage were used at the same time.  In this
case, the custom metric could overwrite the CPU metric, or vice versa.

This fixes that bug, and ensures that the fuzzer tests round-tripping
with multiple metrics.

deads2k commented on pkg/controller/podautoscaler/horizontal_test.go:
```diff
@@ -183,17 +183,39 @@ func (tc *testCase) prepareTestClient(t *testing.T) *fake.Clientset {
-		return true, obj, nil
+
+		// and... convert to autoscaling v1 to return the right type
+		objv1, err := UnsafeConvertToVersionVia(obj, autoscalingv1.SchemeGroupVersion)
```
do a deep copy first.
